src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -221,18 +224,117 @@ impl JoinGraph {
         }
     }
 
+    fn apply_projections_and_filters_to_plan_builder(
+        &self,
+        mut plan_builder: LogicalPlanBuilder,
+    ) -> DaftResult<LogicalPlanBuilder> {
+        // Apply projections and filters in post-traversal order.
+        let mut reversed_items = self.final_projections_and_filters.iter().rev().peekable();
+        while let Some(projection_or_filter) = reversed_items.next() {
+            let is_last = reversed_items.peek().is_none();
+
+            match projection_or_filter {
+                ProjectionOrFilter::Projection(projections) => {
+                    if is_last {
+                        // The final projection is the output projection, so here we select the final projection.
+                        plan_builder = plan_builder.select(projections.clone())?;
```

Comment:
do we have to have all these clones here?

Reply:
We can remove them by `drain`ing the projections and filters instead of iterating

Reply:
Done

Reply:
Ah my bad I did the refactor but forgot the whole point of the refactor. And sounds good, uses pop instead of draining and peeking